File: mobile/flutter/lib/data/models/consistency.dart
import 'package:json_annotation/json_annotation.dart';

part 'consistency.g.dart';

/// Day of week enum (0=Sunday, 6=Saturday - matching backend)
enum DayOfWeek {
  sunday,
  monday,
  tuesday,
  wednesday,
  thursday,
  friday,
  saturday;

  String get displayName {
    switch (this) {
      case DayOfWeek.sunday:
        return 'Sunday';
      case DayOfWeek.monday:
        return 'Monday';
      case DayOfWeek.tuesday:
        return 'Tuesday';
      case DayOfWeek.wednesday:
        return 'Wednesday';
      case DayOfWeek.thursday:
        return 'Thursday';
      case DayOfWeek.friday:
        return 'Friday';
      case DayOfWeek.saturday:
        return 'Saturday';
    }
  }

  String get shortName {
    switch (this) {
      case DayOfWeek.sunday:
        return 'Sun';
      case DayOfWeek.monday:
        return 'Mon';
      case DayOfWeek.tuesday:
        return 'Tue';
      case DayOfWeek.wednesday:
        return 'Wed';
      case DayOfWeek.thursday:
        return 'Thu';
      case DayOfWeek.friday:
        return 'Fri';
      case DayOfWeek.saturday:
        return 'Sat';
    }
  }

  static DayOfWeek fromInt(int value) {
    return DayOfWeek.values[value.clamp(0, 6)];
  }
}

/// Weekly trend direction
enum WeeklyTrend {
  @JsonValue('improving')
  improving,
  @JsonValue('stable')
  stable,
  @JsonValue('declining')
  declining;

  String get displayName {
    switch (this) {
      case WeeklyTrend.improving:
        return 'Improving';
      case WeeklyTrend.stable:
        return 'Stable';
      case WeeklyTrend.declining:
        return 'Declining';
    }
  }
}

/// Calendar heatmap status
enum CalendarStatus {
  @JsonValue('completed')
  completed,
  @JsonValue('missed')
  missed,
  @JsonValue('rest')
  rest,
  @JsonValue('future')
  future;
}

/// Day pattern data for a specific day of the week
@JsonSerializable()
class DayPattern {
  @JsonKey(name: 'day_of_week')
  final int dayOfWeek;

  @JsonKey(name: 'day_name')
  final String dayName;

  @JsonKey(name: 'total_completions')
  final int totalCompletions;

  @JsonKey(name: 'total_skips')
  final int totalSkips;

  @JsonKey(name: 'completion_rate')
  final double completionRate;

  @JsonKey(name: 'is_best_day')
  final bool isBestDay;

  @JsonKey(name: 'is_worst_day')
  final bool isWorstDay;

  const DayPattern({
    required this.dayOfWeek,
    required this.dayName,
    this.totalCompletions = 0,
    this.totalSkips = 0,
    this.completionRate = 0.0,
    this.isBestDay = false,
    this.isWorstDay = false,
  });

  factory DayPattern.fromJson(Map<String, dynamic> json) =>
      _$DayPatternFromJson(json);
  Map<String, dynamic> toJson() => _$DayPatternToJson(this);

  DayOfWeek get dayOfWeekEnum => DayOfWeek.fromInt(dayOfWeek);

  int get totalAttempts => totalCompletions + totalSkips;
}

/// Time of day pattern data
@JsonSerializable()
class TimeOfDayPattern {
  @JsonKey(name: 'time_of_day')
  final String timeOfDay;

  @JsonKey(name: 'display_name')
  final String displayName;

  @JsonKey(name: 'total_completions')
  final int totalCompletions;

  @JsonKey(name: 'total_skips')
  final int totalSkips;

  @JsonKey(name: 'completion_rate')
  final double completionRate;

  @JsonKey(name: 'is_preferred')
  final bool isPreferred;

  const TimeOfDayPattern({
    required this.timeOfDay,
    required this.displayName,
    this.totalCompletions = 0,
    this.totalSkips = 0,
    this.completionRate = 0.0,
    this.isPreferred = false,
  });

  factory TimeOfDayPattern.fromJson(Map<String, dynamic> json) =>
      _$TimeOfDayPatternFromJson(json);
  Map<String, dynamic> toJson() => _$TimeOfDayPatternToJson(this);
}

/// Weekly consistency metric
@JsonSerializable()
class WeeklyConsistencyMetric {
  @JsonKey(name: 'week_start')
  final String weekStart;

  @JsonKey(name: 'week_end')
  final String weekEnd;

  @JsonKey(name: 'workouts_scheduled')
  final int workoutsScheduled;

  @JsonKey(name: 'workouts_completed')
  final int workoutsCompleted;

  @JsonKey(name: 'completion_rate')
  final double completionRate;

  @JsonKey(name: 'total_workout_minutes')
  final int totalWorkoutMinutes;

  const WeeklyConsistencyMetric({
    required this.weekStart,
    required this.weekEnd,
    this.workoutsScheduled = 0,
    this.workoutsCompleted = 0,
    this.completionRate = 0.0,
    this.totalWorkoutMinutes = 0,
  });

  factory WeeklyConsistencyMetric.fromJson(Map<String, dynamic> json) =>
      _$WeeklyConsistencyMetricFromJson(json);
  Map<String, dynamic> toJson() => _$WeeklyConsistencyMetricToJson(this);

  DateTime get weekStartDate => DateTime.parse(weekStart);
  DateTime get weekEndDate => DateTime.parse(weekEnd);

  int get workoutsMissed => workoutsScheduled - workoutsCompleted;
}

/// Comprehensive consistency insights
@JsonSerializable()
class ConsistencyInsights {
  @JsonKey(name: 'user_id')
  final String userId;

  // Current state
  @JsonKey(name: 'current_streak')
  final int currentStreak;

  @JsonKey(name: 'longest_streak')
  final int longestStreak;

  @JsonKey(name: 'is_streak_active')
  final bool isStreakActive;

  // Day patterns
  @JsonKey(name: 'best_day')
  final DayPattern? bestDay;

  @JsonKey(name: 'worst_day')
  final DayPattern? worstDay;

  @JsonKey(name: 'day_patterns')
  final List<DayPattern> dayPatterns;

  // Time preferences
  @JsonKey(name: 'preferred_time')
  final String? preferredTime;

  @JsonKey(name: 'time_patterns')
  final List<TimeOfDayPattern> timePatterns;

  // Monthly stats
  @JsonKey(name: 'month_workouts_completed')
  final int monthWorkoutsCompleted;

  @JsonKey(name: 'month_workouts_scheduled')
  final int monthWorkoutsScheduled;

  @JsonKey(name: 'month_completion_rate')
  final double monthCompletionRate;

  @JsonKey(name: 'month_display')
  final String monthDisplay;

  // Weekly stats
  @JsonKey(name: 'weekly_completion_rates')
  final List<WeeklyConsistencyMetric> weeklyCompletionRates;

  @JsonKey(name: 'average_weekly_rate')
  final double averageWeeklyRate;

  @JsonKey(name: 'weekly_trend')
  final String weeklyTrend;

  // Recovery
  @JsonKey(name: 'needs_recovery')
  final bool needsRecovery;

  @JsonKey(name: 'recovery_suggestion')
  final String? recoverySuggestion;

  @JsonKey(name: 'days_since_last_workout')
  final int daysSinceLastWorkout;

  @JsonKey(name: 'last_workout_date')
  final String? lastWorkoutDate;

  @JsonKey(name: 'calculated_at')
  final String? calculatedAt;

  const ConsistencyInsights({
    required this.userId,
    this.currentStreak = 0,
    this.longestStreak = 0,
    this.isStreakActive = false,
    this.bestDay,
    this.worstDay,
    this.dayPatterns = const [],
    this.preferredTime,
    this.timePatterns = const [],
    this.monthWorkoutsCompleted = 0,
    this.monthWorkoutsScheduled = 0,
    this.monthCompletionRate = 0.0,
    this.monthDisplay = '',
    this.weeklyCompletionRates = const [],
    this.averageWeeklyRate = 0.0,
    this.weeklyTrend = 'stable',
    this.needsRecovery = false,
    this.recoverySuggestion,
    this.daysSinceLastWorkout = 0,
    this.lastWorkoutDate,
    this.calculatedAt,
  });

  factory ConsistencyInsights.fromJson(Map<String, dynamic> json) =>
      _$ConsistencyInsightsFromJson(json);
  Map<String, dynamic> toJson() => _$ConsistencyInsightsToJson(this);

  WeeklyTrend get weeklyTrendEnum {
    switch (weeklyTrend.toLowerCase()) {
      case 'improving':
        return WeeklyTrend.improving;
      case 'declining':
        return WeeklyTrend.declining;
      default:
        return WeeklyTrend.stable;
    }
  }

  DateTime? get lastWorkoutDateTime =>
      lastWorkoutDate != null ? DateTime.tryParse(lastWorkoutDate!) : null;

  /// Get a motivational message based on streak status
  String get streakMessage {
    if (currentStreak == 0) {
      if (needsRecovery) {
        return 'Start fresh today!';
      }
      return 'Begin your streak today!';
    } else if (currentStreak == 1) {
      return 'Day 1 - Great start!';
    } else if (currentStreak < 7) {
      return '$currentStreak days strong!';
    } else if (currentStreak < 30) {
      return '$currentStreak day streak!';
    } else {
      return '$currentStreak days - Incredible!';
    }
  }

  /// Get best day display string
  String? get bestDayDisplay {
    if (bestDay == null) return null;
    return '${bestDay!.dayName} (${bestDay!.completionRate.toStringAsFixed(0)}%)';
  }

  /// Get worst day display string
  String? get worstDayDisplay {
    if (worstDay == null) return null;
    return '${worstDay!.dayName} (${worstDay!.completionRate.toStringAsFixed(0)}%)';
  }
}

/// Streak history record
@JsonSerializable()
class StreakHistoryRecord {
  final String id;

  @JsonKey(name: 'user_id')
  final String userId;

  @JsonKey(name: 'streak_length')
  final int streakLength;

  @JsonKey(name: 'started_at')
  final String startedAt;

  @JsonKey(name: 'ended_at')
  final String endedAt;

  @JsonKey(name: 'end_reason')
  final String? endReason;

  @JsonKey(name: 'created_at')
  final String createdAt;

  const StreakHistoryRecord({
    required this.id,
    required this.userId,
    required this.streakLength,
    required this.startedAt,
    required this.endedAt,
    this.endReason,
    required this.createdAt,
  });

  factory StreakHistoryRecord.fromJson(Map<String, dynamic> json) =>
      _$StreakHistoryRecordFromJson(json);
  Map<String, dynamic> toJson() => _$StreakHistoryRecordToJson(this);

  DateTime get startedAtDate => DateTime.parse(startedAt);
  DateTime get endedAtDate => DateTime.parse(endedAt);
}

/// Detailed consistency patterns
@JsonSerializable()
class ConsistencyPatterns {
  @JsonKey(name: 'user_id')
  final String userId;

  @JsonKey(name: 'time_patterns')
  final List<TimeOfDayPattern> timePatterns;

  @JsonKey(name: 'preferred_time')
  final String? preferredTime;

  @JsonKey(name: 'day_patterns')
  final List<DayPattern> dayPatterns;

  @JsonKey(name: 'most_consistent_day')
  final String? mostConsistentDay;

  @JsonKey(name: 'least_consistent_day')
  final String? leastConsistentDay;

  @JsonKey(name: 'has_seasonal_data')
  final bool hasSeasonalData;

  @JsonKey(name: 'seasonal_notes')
  final String? seasonalNotes;

  @JsonKey(name: 'skip_reasons')
  final Map<String, int> skipReasons;

  @JsonKey(name: 'most_common_skip_reason')
  final String? mostCommonSkipReason;

  @JsonKey(name: 'streak_history')
  final List<StreakHistoryRecord> streakHistory;

  @JsonKey(name: 'average_streak_length')
  final double averageStreakLength;

  @JsonKey(name: 'streak_count')
  final int streakCount;

  @JsonKey(name: 'calculated_at')
  final String? calculatedAt;

  const ConsistencyPatterns({
    required this.userId,
    this.timePatterns = const [],
    this.preferredTime,
    this.dayPatterns = const [],
    this.mostConsistentDay,
    this.leastConsistentDay,
    this.hasSeasonalData = false,
    this.seasonalNotes,
    this.skipReasons = const {},
    this.mostCommonSkipReason,
    this.streakHistory = const [],
    this.averageStreakLength = 0.0,
    this.streakCount = 0,
    this.calculatedAt,
  });

  factory ConsistencyPatterns.fromJson(Map<String, dynamic> json) =>
      _$ConsistencyPatternsFromJson(json);
  Map<String, dynamic> toJson() => _$ConsistencyPatternsToJson(this);
}

/// Calendar heatmap data point
@JsonSerializable()
class CalendarHeatmapData {
  final String date;

  @JsonKey(name: 'day_of_week')
  final int dayOfWeek;

  final String status;

  @JsonKey(name: 'workout_name')
  final String? workoutName;

  const CalendarHeatmapData({
    required this.date,
    required this.dayOfWeek,
    required this.status,
    this.workoutName,
  });

  factory CalendarHeatmapData.fromJson(Map<String, dynamic> json) =>
      _$CalendarHeatmapDataFromJson(json);
  Map<String, dynamic> toJson() => _$CalendarHeatmapDataToJson(this);

  DateTime get dateTime => DateTime.parse(date);

  CalendarStatus get statusEnum {
    switch (status.toLowerCase()) {
      case 'completed':
        return CalendarStatus.completed;
      case 'missed':
        return CalendarStatus.missed;
      case 'future':
        return CalendarStatus.future;
      default:
        return CalendarStatus.rest;
    }
  }
}

/// Calendar heatmap response
@JsonSerializable()
class CalendarHeatmapResponse {
  @JsonKey(name: 'user_id')
  final String userId;

  @JsonKey(name: 'start_date')
  final String startDate;

  @JsonKey(name: 'end_date')
  final String endDate;

  final List<CalendarHeatmapData> data;

  @JsonKey(name: 'total_completed')
  final int totalCompleted;

  @JsonKey(name: 'total_missed')
  final int totalMissed;

  @JsonKey(name: 'total_rest_days')
  final int totalRestDays;

  const CalendarHeatmapResponse({
    required this.userId,
    required this.startDate,
    required this.endDate,
    this.data = const [],
    this.totalCompleted = 0,
    this.totalMissed = 0,
    this.totalRestDays = 0,
  });

  factory CalendarHeatmapResponse.fromJson(Map<String, dynamic> json) =>
      _$CalendarHeatmapResponseFromJson(json);
  Map<String, dynamic> toJson() => _$CalendarHeatmapResponseToJson(this);

  DateTime get startDateTime => DateTime.parse(startDate);
  DateTime get endDateTime => DateTime.parse(endDate);

  int get totalDays => data.length;
}

/// Streak recovery response
@JsonSerializable()
class StreakRecoveryResponse {
  final bool success;

  @JsonKey(name: 'attempt_id')
  final String attemptId;

  final String message;

  @JsonKey(name: 'motivation_quote')
  final String? motivationQuote;

  @JsonKey(name: 'suggested_workout_type')
  final String? suggestedWorkoutType;

  @JsonKey(name: 'suggested_duration_minutes')
  final int? suggestedDurationMinutes;

  const StreakRecoveryResponse({
    required this.success,
    required this.attemptId,
    required this.message,
    this.motivationQuote,
    this.suggestedWorkoutType,
    this.suggestedDurationMinutes,
  });

  factory StreakRecoveryResponse.fromJson(Map<String, dynamic> json) =>
      _$StreakRecoveryResponseFromJson(json);
  Map<String, dynamic> toJson() => _$StreakRecoveryResponseToJson(this);
}

File: backend/models/consistency.py
"""
Consistency Analytics Models
============================
Pydantic models for consistency tracking, streak analysis, and workout patterns.
"""

from datetime import datetime, date
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from enum import Enum


# ============================================================================
# Enums
# ============================================================================

class DayOfWeek(int, Enum):
    """Day of week (0=Sunday, 6=Saturday - matching SQL EXTRACT(DOW))."""
    SUNDAY = 0
    MONDAY = 1
    TUESDAY = 2
    WEDNESDAY = 3
    THURSDAY = 4
    FRIDAY = 5
    SATURDAY = 6

    @property
    def display_name(self) -> str:
        """Get human-readable name."""
        names = {
            0: "Sunday",
            1: "Monday",
            2: "Tuesday",
            3: "Wednesday",
            4: "Thursday",
            5: "Friday",
            6: "Saturday",
        }
        return names.get(self.value, "Unknown")

    @property
    def short_name(self) -> str:
        """Get abbreviated name."""
        names = {
            0: "Sun",
            1: "Mon",
            2: "Tue",
            3: "Wed",
            4: "Thu",
            5: "Fri",
            6: "Sat",
        }
        return names.get(self.value, "?")


class TimeOfDay(str, Enum):
    """Time of day preference for workouts."""
    EARLY_MORNING = "early_morning"  # 5-8 AM
    MORNING = "morning"  # 8-11 AM
    MIDDAY = "midday"  # 11 AM - 2 PM
    AFTERNOON = "afternoon"  # 2-5 PM
    EVENING = "evening"  # 5-8 PM
    NIGHT = "night"  # 8-11 PM

    @classmethod
    def from_hour(cls, hour: int) -> "TimeOfDay":
        """Convert hour (0-23) to TimeOfDay."""
        if 5 <= hour < 8:
            return cls.EARLY_MORNING
        elif 8 <= hour < 11:
            return cls.MORNING
        elif 11 <= hour < 14:
            return cls.MIDDAY
        elif 14 <= hour < 17:
            return cls.AFTERNOON
        elif 17 <= hour < 20:
            return cls.EVENING
        else:
            return cls.NIGHT


class StreakEndReason(str, Enum):
    """Reason why a streak ended."""
    MISSED_WORKOUT = "missed_workout"
    MANUAL_RESET = "manual_reset"
    PROGRAM_CHANGE = "program_change"
    ACCOUNT_ISSUE = "account_issue"


class RecoveryType(str, Enum):
    """Type of streak recovery workout."""
    STANDARD = "standard"
    QUICK_RECOVERY = "quick_recovery"  # Shorter, easier workout
    CUSTOM = "custom"


# ============================================================================
# Streak Models
# ============================================================================

class StreakHistoryRecord(BaseModel):
    """A historical streak record."""
    id: str
    user_id: str
    streak_length: int = Field(..., ge=0)
    started_at: datetime
    ended_at: datetime
    end_reason: Optional[str] = StreakEndReason.MISSED_WORKOUT.value
    created_at: datetime


class StreakRecoveryAttempt(BaseModel):
    """A streak recovery attempt record."""
    id: str
    user_id: str
    previous_streak_length: int = Field(default=0, ge=0)
    days_since_last_workout: int = Field(default=1, ge=0)
    recovery_workout_id: Optional[str] = None
    recovery_type: str = RecoveryType.STANDARD.value
    motivation_message: Optional[str] = None
    was_successful: Optional[bool] = None
    created_at: datetime
    completed_at: Optional[datetime] = None


class StreakRecoveryRequest(BaseModel):
    """Request to start a streak recovery."""
    user_id: str
    recovery_type: str = Field(default=RecoveryType.STANDARD.value)


class StreakRecoveryResponse(BaseModel):
    """Response after initiating streak recovery."""
    success: bool
    attempt_id: str
    message: str
    motivation_quote: Optional[str] = None
    suggested_workout_type: Optional[str] = None
    suggested_duration_minutes: Optional[int] = None


# ============================================================================
# Time Pattern Models
# ============================================================================

class WorkoutTimePattern(BaseModel):
    """Workout completion pattern for a specific day/hour."""
    id: Optional[str] = None
    user_id: str
    day_of_week: int = Field(..., ge=0, le=6)
    hour_of_day: int = Field(..., ge=0, le=23)
    completion_count: int = Field(default=0, ge=0)
    skip_count: int = Field(default=0, ge=0)
    updated_at: Optional[datetime] = None

    @property
    def total_attempts(self) -> int:
        """Total scheduled workouts at this time."""
        return self.completion_count + self.skip_count

    @property
    def completion_rate(self) -> float:
        """Completion rate as percentage (0-100)."""
        if self.total_attempts == 0:
            return 0.0
        return round((self.completion_count / self.total_attempts) * 100, 1)

    @property
    def day_name(self) -> str:
        """Human-readable day name."""
        return DayOfWeek(self.day_of_week).display_name

    @property
    def time_of_day(self) -> TimeOfDay:
        """Time of day category."""
        return TimeOfDay.from_hour(self.hour_of_day)


class DayPattern(BaseModel):
    """Aggregated pattern for a single day of the week."""
    day_of_week: int = Field(..., ge=0, le=6)
    day_name: str
    total_completions: int = 0
    total_skips: int = 0
    completion_rate: float = 0.0
    is_best_day: bool = False
    is_worst_day: bool = False


class TimeOfDayPattern(BaseModel):
    """Aggregated pattern for a time of day."""
    time_of_day: str
    display_name: str
    total_completions: int = 0
    total_skips: int = 0
    completion_rate: float = 0.0
    is_preferred: bool = False


# ============================================================================
# Daily Metrics Models
# ============================================================================

class DailyConsistencyMetric(BaseModel):
    """Daily consistency metrics record."""
    id: Optional[str] = None
    user_id: str
    metric_date: date
    workouts_scheduled: int = Field(default=0, ge=0)
    workouts_completed: int = Field(default=0, ge=0)
    workouts_skipped: int = Field(default=0, ge=0)
    total_workout_minutes: int = Field(default=0, ge=0)
    streak_day: int = Field(default=0, ge=0)
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    @property
    def completion_rate(self) -> float:
        """Daily completion rate."""
        if self.workouts_scheduled == 0:
            return 100.0 if self.workouts_completed > 0 else 0.0
        return round((self.workouts_completed / self.workouts_scheduled) * 100, 1)


class WeeklyConsistencyMetric(BaseModel):
    """Aggregated weekly consistency metrics."""
    week_start: date
    week_end: date
    workouts_scheduled: int = 0
    workouts_completed: int = 0
    workouts_skipped: int = 0
    completion_rate: float = 0.0
    total_workout_minutes: int = 0
    average_session_minutes: float = 0.0


# ============================================================================
# Insights Response Models
# ============================================================================

class ConsistencyInsights(BaseModel):
    """Comprehensive consistency insights for a user."""
    user_id: str

    # Current state
    current_streak: int = Field(default=0, ge=0)
    longest_streak: int = Field(default=0, ge=0)
    is_streak_active: bool = False

    # Day patterns
    best_day: Optional[DayPattern] = None
    worst_day: Optional[DayPattern] = None
    day_patterns: List[DayPattern] = Field(default_factory=list)

    # Time preferences
    preferred_time: Optional[str] = None  # TimeOfDay value
    time_patterns: List[TimeOfDayPattern] = Field(default_factory=list)

    # Monthly stats
    month_workouts_completed: int = 0
    month_workouts_scheduled: int = 0
    month_completion_rate: float = 0.0
    month_display: str = ""  # "12 of 16 workouts"

    # Weekly stats (last 4 weeks)
    weekly_completion_rates: List[WeeklyConsistencyMetric] = Field(default_factory=list)
    average_weekly_rate: float = 0.0
    weekly_trend: str = "stable"  # improving, stable, declining

    # Streak recovery suggestion (if streak broken)
    needs_recovery: bool = False
    recovery_suggestion: Optional[str] = None
    days_since_last_workout: int = 0

    # Computed timestamps
    last_workout_date: Optional[date] = None
    calculated_at: datetime = Field(default_factory=datetime.now)


class ConsistencyPatterns(BaseModel):
    """Detailed consistency patterns analysis."""
    user_id: str

    # Time of day patterns
    time_patterns: List[TimeOfDayPattern] = Field(default_factory=list)
    preferred_time: Optional[str] = None

    # Day of week patterns
    day_patterns: List[DayPattern] = Field(default_factory=list)
    most_consistent_day: Optional[str] = None
    least_consistent_day: Optional[str] = None

    # Seasonal patterns (if enough data)
    has_seasonal_data: bool = False
    seasonal_notes: Optional[str] = None

    # Skip reasons breakdown (if tracked)
    skip_reasons: Dict[str, int] = Field(default_factory=dict)
    most_common_skip_reason: Optional[str] = None

    # Historical streaks
    streak_history: List[StreakHistoryRecord] = Field(default_factory=list)
    average_streak_length: float = 0.0
    streak_count: int = 0

    calculated_at: datetime = Field(default_factory=datetime.now)


# ============================================================================
# Request/Response Models for API
# ============================================================================

class ConsistencyInsightsRequest(BaseModel):
    """Request parameters for consistency insights."""
    user_id: str
    include_patterns: bool = Field(default=False)
    days_back: int = Field(default=90, ge=7, le=365)


class ConsistencyPatternsRequest(BaseModel):
    """Request parameters for detailed patterns."""
    user_id: str
    days_back: int = Field(default=180, ge=30, le=365)


class RecordWorkoutPatternRequest(BaseModel):
    """Request to record a workout pattern update."""
    user_id: str
    completed_at: datetime
    is_completed: bool


class CalendarHeatmapData(BaseModel):
    """Data for the weekly calendar heatmap visualization."""
    date: date
    day_of_week: int
    status: str  # "completed", "missed", "rest", "future"
    workout_name: Optional[str] = None


class CalendarHeatmapResponse(BaseModel):
    """Response containing calendar heatmap data."""
    user_id: str
    start_date: date
    end_date: date
    data: List[CalendarHeatmapData] = Field(default_factory=list)
    total_completed: int = 0
    total_missed: int = 0
    total_rest_days: int = 0
